First push to repo of dmd hospital only

brianmackenna opened the pull request:
This is the first push of the notebook to this repo. I was working on it in Jupyter Notebook bucket but as per manifesto and best practice I am establishing a singular notebook for further work on this. The intention presently is to publish a blog as this raises some important issues e.g. safety, accuracy/updates of dm+d etc. 

@inglesp  would it be possible to get a (quick) code review pleaseas per the manifesto? Particular area to check

- [x] Input box 2: First SQL. 
---- is the SQL correct for what you think I am attempting to do? (see box below)
---- am I using the correct dmd tables in BQ? (this is why I asked you specifically for code review)

- [x] Ouput box 5: Some duplicate names. I am getting duplicate `bnf_nam` and `bnf_code`. I'm not sure what is happening but I suspect it is to do with mapping dmd elements to bnf names and code and is related to the SQL above. Do you have a better understanding of  what is causing it?

- [x] Output box 17: I have created csv files for CCGs to investigate further. It would be neater to create these in a separate folder in the repo but I can't seem to do it. Do you know off the top of your head how to send these csv to a newly created folder in the repo?

brianmackenna committed:
First push to repo

This is the first push of the notebook to this repo. I was working on it in Jupyter Notebook bucket but as per manifesto and best practice I am establishing a singular notebook for further work on this.

inglesp commented:
### Input box 2: First SQL.

You're missing out a bunch of BNF codes where the VMP and AMP have different BNF codes, for instance [Methyl aminolevulinate 16% cream](https://openprescribing.net/dmd/vmp/408157009/relationships/).

![image](https://user-images.githubusercontent.com/28734/71989878-5851ea00-322a-11ea-9c2e-d8998e313637.png)

In this case, there is prescribing for both the generic and the branded BNF codes:

```
SELECT
  bnf_code,
  SUM(items) AS total_items
FROM hscic.normalised_prescribing_standard
WHERE bnf_code IN ('1308010X0AAAAAA', '1308010X0BBAAAA')
GROUP BY bnf_code
```


Row | bnf_code | total_items
-- | -- | --
1 | 1308010X0BBAAAA | 21
2 | 1308010X0AAAAAA | 17

So you're going to be missing several things from your analysis.

I suggest the correct fix is to create a view similar to `dmd_objs_with_form_route` that allows us to look up all BNF codes for each availability restriction.  This would also shed light onto whether availability restrictions are sensibly defined at BNF code level -- you could imagine a situation where some of a VMP's AMPs were hospital only and some weren't.

I can create this later on today.

Longer term (and especially if we don't get SNOMED-level DPI data from the BSA) we should attempt to map from BNF code to each dm+d property.

inglesp commented:
### Input box 2: First SQL

> is the SQL correct for what you think I am attempting to do?

Not quite, see below.

> am I using the correct dmd tables in BQ?

Yes, as far as I can tell.

---

I have done some more digging into how we can identify presentations that should be hospital only.  It's a bit fiddly, and there are some ambiguities.

We are currently just considering presentations that map to at least one AMP which is hospital only.  But some presentations map both to AMPs which are hospital only and to some which are not.

<details>
<summary>SQL</summary>

```
SELECT hospital_only, COUNT(*) AS c
FROM (
  SELECT
    bnf_code,
    CASE COUNT(CASE WHEN avail_restrict = 8 THEN 1 ELSE NULL END)
    WHEN 0 THEN "Not hospital only"
    WHEN COUNT(*) THEN "Hospital only"
    ELSE "Mixture"
    END AS hospital_only
  FROM
    dmd.amp
  WHERE
    avail_restrict != 9 AND
    bnf_code IS NOT NULL
  GROUP BY
    bnf_code
  ORDER BY
    bnf_code
)
GROUP BY hospital_only
```

</details>

Row | hospital_only | c
-- | -- | --
1 | Not hospital only | 31904
2 | Mixture | 104
3 | Hospital only | 471

An example of a presentation which has a mixture is Pantoprazole_Inj 40mg Vl (0103050R0AAACAC).  The AMPs produced by [AAH](https://openprescribing.net/dmd/amp/18276611000001102/) and [Bowmed Ibisqus](https://openprescribing.net/dmd/amp/34755011000001107/) are hospital only, and those produced by [Reig Jofre](https://openprescribing.net/dmd/amp/36868311000001103/) and [Sun](https://openprescribing.net/dmd/amp/21507211000001106/) are not.  There is [some prescribing](https://openprescribing.net/analyse/#org=CCG&numIds=0103050R0AAACAC&denom=nothing&selectedTab=chart) for this presentation (tens of items/month) but we have no way of knowing whether hospital only AMPs were prescribed.

This affects 104 / 575 (=18%) of the presentations that are picked up by the query in input box 2, and we'll need to decide whether to include these in the analysis (as we currently do) or not.

---

There are also some presentations that map only to VMPs, and we should consider those too.  Taking these into account:


<details>
<summary>SQL</summary>

```
WITH s AS (
  SELECT bnf_code, avail_restrict
  FROM dmd.amp
  WHERE avail_restrict != 9 AND bnf_code IS NOT NULL

  UNION DISTINCT

  SELECT vmp.bnf_code, amp.avail_restrict
  FROM dmd.vmp
  INNER JOIN dmd.amp
    ON vmp.id = amp.vmp
  WHERE amp.avail_restrict != 9 AND vmp.bnf_code IS NOT NULL

  ORDER BY bnf_code
)

SELECT hospital_only, COUNT(*) AS c
FROM (
  SELECT
    bnf_code,
    CASE COUNT(CASE WHEN avail_restrict = 8 THEN 1 ELSE NULL END)
    WHEN 0 THEN "Not hospital only"
    WHEN COUNT(*) THEN "Hospital only"
    ELSE "Mixture"
    END AS hospital_only
  FROM
    s
  WHERE
    avail_restrict != 9 AND
    bnf_code IS NOT NULL
  GROUP BY
    bnf_code
  ORDER BY
    bnf_code
)
GROUP BY hospital_only
```

</details>

Row | hospital_only | c |  
-- | -- | --
1 | Not hospital only | 34911
2 | Mixture | 179
3 | Hospital only | 684

So there are 213 extra hospital only presentations that can be found by including presentations that map only to VMPs.

---

So what should the SQL for input 2 look like?  There are two cases we might want: (a) we want prescribing for all presentations where at least one AMP is hospital only, or (b) we want prescribing for all presentations where all AMPs are hospital only.

I'm going to refactor the SQL in input 2 so that we isolate the bit where we work out which BNF codes we're interested in.

For (a):

```
WITH presentations_with_some_hospital_only_amps AS (
  SELECT bnf_code
  FROM dmd.amp
  WHERE avail_restrict = 8 AND bnf_code IS NOT NULL
  
  UNION DISTINCT
  
  SELECT vmp.bnf_code
  FROM dmd.vmp
  INNER JOIN dmd.amp
    ON vmp.id = amp.vmp
  WHERE amp.avail_restrict = 8 AND vmp.bnf_code IS NOT NULL
)

SELECT
  pct,
  month,
  presentation AS bnf_name,
  bnf_code,
  SUM(items) AS total_items,
  SUM(actual_cost) AS total_cost
FROM
  hscic.normalised_prescribing_standard AS rx
INNER JOIN
  hscic.ccgs
ON
  rx.pct = ccgs.code
INNER JOIN
  hscic.bnf
ON
  rx.bnf_code = bnf.presentation_code
WHERE
  org_type = 'CCG'
  AND bnf_code IN (SELECT bnf_code FROM presentations_with_some_hospital_only_amps)
GROUP BY
  pct,
  month,
  bnf_code,
  bnf_name
ORDER BY
  pct,
  month,
  bnf_code
```

And for (b):

```
WITH bnf_codes_and_avail_restricts AS (
  SELECT bnf_code, avail_restrict
  FROM dmd.amp
  WHERE avail_restrict != 9 AND bnf_code IS NOT NULL
  
  UNION DISTINCT
  
  SELECT vmp.bnf_code, amp.avail_restrict
  FROM dmd.vmp
  INNER JOIN dmd.amp
    ON vmp.id = amp.vmp
  WHERE amp.avail_restrict != 9 AND vmp.bnf_code IS NOT NULL
),

presentations_with_only_hospital_only_amps AS (
  SELECT bnf_code
  FROM bnf_codes_and_avail_restricts
  GROUP BY bnf_code
  HAVING COUNT(CASE WHEN avail_restrict = 8 THEN 1 ELSE NULL END) = COUNT(*)
)

SELECT
  pct,
  month,
  presentation AS bnf_name,
  bnf_code,
  SUM(items) AS total_items,
  SUM(actual_cost) AS total_cost
FROM
  hscic.normalised_prescribing_standard AS rx
INNER JOIN
  hscic.ccgs
ON
  rx.pct = ccgs.code
INNER JOIN
  hscic.bnf
ON rx.bnf_code = bnf.presentation_code
WHERE
  org_type = 'CCG'
  AND bnf_code IN (SELECT bnf_code FROM presentations_with_only_hospital_only_amps)
GROUP BY
  pct,
  month,
  bnf_code,
  bnf_name
ORDER BY
  pct,
  month,
  bnf_code
```

---

The main difference with your SQL (apart from finding all the presentations correctly) is that I've used `WHERE bnf_code IN (SELECT ...)` instead of a `JOIN` to filter the `normalised_prescribing_standard` table.  A `JOIN` gives the correct results, but I would only expect to see a `JOIN` used to pull in information from another table, and not for filtering.

I've also `JOIN`ed on the `bnf` table to pull in the correct BNF name, rather than using the BNF name that was correct at the time the data was published.  (See below for why this is important.)

And a couple of minor things:

* I've removed the `AS ccgs` table alias, because the table is already called `ccgs`
* I've removed the table qualifiers (`rx.` and `ccgs.`) from everywhere except in the `ON` clause.  They're not necessary, because there's no ambiguity about which table each field belongs to.  (They're not even necessary in the `ON` clause, but it looks odd to me without them)  You could argue that the table qualifiers make the SQL more explicit and so easier to understand, but they should be used consistently.
* I've renamed `JOIN` to `INNER JOIN` for clarity.
* I've added an `ORDER BY` to ensure the results always come out in the same order.  I find this helps when I'm comparing two queries that should return similar things.

---

### Ouput box 5: Some duplicate names

It looks like there are duplicate names, but actually there are some sneaky invisible space characters confusing things:

![image](https://user-images.githubusercontent.com/28734/72160811-e064fa80-33b6-11ea-900c-7c0e00d5e88a.png)

This must be because `hscic.normalised_prescribing_standard` contains entries with `bnf_name`s that end in a space.

<details>
<summary>In fact more than half of the `bnf_name`s end in a space, yikes</summary>

```
SELECT n, total, n / total AS proportion
FROM (
  SELECT
    (SELECT COUNT(*) FROM hscic.normalised_prescribing_standard WHERE bnf_name LIKE '% ') AS n,
    (SELECT COUNT(*) FROM hscic.normalised_prescribing_standard) AS total
)
```


Row | n | total | proportion |  
-- | -- | -- | -- | --
1 | 693019926 | 1103381540 | 0.6280872942645026 |  

</details>

This is fixed by my SQL above, because rather than reading the BNF name from `hscic.normalised_prescribing_standard`, we `JOIN` on the `hscic.bnf` table which has the canonical names for each BNF code.  Even if `hscic.normalised_prescribing_standard` didn't have stray trailing spaces, this would be good practice because BNF names can and do change, and we want to make sure that we only use the current name.

---

### Output box 17: Separate folders for CSV files

Quick detour:

I'm not a pandas expert, so without running the code, I don't know what `i` and `g` are here:

```
for i, g in latest_hosp_df_1000.groupby('pct'):
     g.to_csv('{}.csv'.format(i.split('/')[0]), index=False)
```

Could you give them better names?  `i` looks like it's the CCG code, so you could rename it `ccg_id` or `ccg_code` or `ccg`, while `g` is going to be a dataframe, so you could call it `ccg_df`.

```
for ccg_id, ccg_df in latest_hosp_df_1000.groupby('pct'):
     ccg_df.to_csv('{}.csv'.format(ccg_id.split('/')[0]), index=False)
```

---

If I've understood this right, then `ccg_id.split('/')[0]` will always be the same as `ccg_id`, since `ccg_id` will never have a `/` in it.

I think what you want is:

```
for ccg_id, ccg_df in latest_hosp_df_1000.groupby('pct'):
     ccg_df.to_csv('ccg-data/{}.csv'.format(ccg_id), index=False)
```

(You might need to change the `/` to a `\` on Windows, I can't remember.)

You'll probably need to make sure that the `ccg-data` directory already exists.

brianmackenna committed:
Post code review changes

Amnded notebook based on feedback from @inglesp
- SQL amended to ientify all products where at least 1 amp is hospital only
- "Mixture" products
- Plotted new graphs
- Created new files and folder for CCGs to investigate locally

To do
EHR impact

brianmackenna committed:
EHR update

added in EHR data

brianmackenna committed:
ehr update

added in ehr data - last attempt broke

brianmackenna commented:
Thanks @inglesp  for a detailed review - I learnt a lot from the feedback and changing it myself.

SQL - I have investigated _(a) we want prescribing for all presentations where at least one AMP is hospital only_ I have also pulled out the "mixture" file to have a closer look. Some are as expected and some are likely data errors or inconsistencies in decisions by companies on what constitutes "hospital only".

Separate folders - this worked thank you. I was going wrong by not creating the folder beforehand and was trying to get pandas to create it. I have found some information [here](https://www.youtube.com/watch?v=cdHgGWtlPEc&feature=youtu.be) and [here](https://gist.github.com/keithweaver/562d3caa8650eefe7f84fa074e9ca949) for my own future reference.

Next step is to share with team for feedback. Possible outputs from this analysis
 - feedback to BSA about the findings particularly products are really not "hospital only" and most likely data entry error
- share with interested people outside team briefly e.g. NHSI safety team, Callum Polwart for feedback
- ship blog asap on findings
- tool/measure for the website: we will need to carefully filter for noise of data errors, antibiotics, "mixture" products.
- Paper - this links with outlier detection work as it is low numbers and some things e.g. zuclopenthixol might be suggestive of a medication error e.g. [zuclopenthixol](https://ebmdatalab.net/zuclopenthixol-acetate-a-new-kind-of-measure-on-openprescribing/) which is now "hospital only" after our prior work
